src/toplevel.c

```diff
@@ -585,8 +592,13 @@ jl_value_t *jl_parse_eval_all(const char *fname, size_t len)
             jl_rethrow();
         }
         else {
-            jl_rethrow_other(jl_new_struct(jl_loaderror_type, fn, ln,
-                                           jl_exception_in_transit));
+            if(jl_typeis(jl_exception_in_transit, jl_initerror_type)) {
```

Comment:
Just double-wrapping the exception is probably better. That way the decision to have InitError is orthogonal to how everything else works. I'm not sure what you can do by catching InitErrors, but at least they provide more information for printing exceptions, which seems harmless enough for now.